File: backend/app/services/task_service.py
"""Task service."""

from app.models import Task, TaskCreate, TaskStatus, TaskStatusUpdate, TaskUpdate, TaskByStatus
from app.repositories import TaskRepository


class TaskService:
    """Service for task operations."""

    def __init__(self):
        self.task_repo = TaskRepository()

    def get_all(self) -> list[Task]:
        """Get all tasks."""
        return self.task_repo.find_all()

    def get_by_id(self, task_id: int) -> Task | None:
        """Get task by ID."""
        return self.task_repo.find_by_id(task_id)

    def get_by_project(self, project_id: int) -> list[Task]:
        """Get all tasks for a project."""
        return self.task_repo.find_by_project(project_id)

    def get_by_project_grouped(self, project_id: int) -> TaskByStatus:
        """Get tasks for a project grouped by status."""
        grouped = self.task_repo.group_by_status(project_id)
        return TaskByStatus(**{k: v for k, v in grouped.items()})

    def create(self, task_data: TaskCreate) -> Task:
        """Create a new task."""
        return self.task_repo.create(task_data)

    def update(self, task_id: int, task_data: TaskUpdate) -> Task | None:
        """Update a task."""
        return self.task_repo.update(task_id, task_data)

    def update_status(self, task_id: int, status_data: TaskStatusUpdate) -> Task | None:
        """Update task status."""
        task = self.task_repo.find_by_id(task_id)
        if not task:
            return None

        update_data = TaskUpdate(status=status_data.status)
        return self.task_repo.update(task_id, update_data)

    def delete(self, task_id: int) -> bool:
        """Delete a task."""
        return self.task_repo.delete(task_id)

    def exists(self, task_id: int) -> bool:
        """Check if task exists."""
        return self.task_repo.find_by_id(task_id) is not None

File: backend-nodejs/src/services/taskService.js
/**
 * Task Service
 */
const { TaskRepository } = require('../repositories/taskRepository');
const { ProjectRepository } = require('../repositories/projectRepository');

class TaskService {
  constructor() {
    this.taskRepo = new TaskRepository();
    this.projectRepo = new ProjectRepository();
  }

  /**
   * Get all tasks
   * @returns {Promise<Array>} All tasks
   */
  async getAll() {
    return await this.taskRepo.findAll();
  }

  /**
   * Get task by ID
   * @param {number} taskId - Task ID
   * @returns {Promise<object|null>} Task or null
   */
  async getById(taskId) {
    return await this.taskRepo.findById(taskId);
  }

  /**
   * Get all tasks for a project
   * @param {number} projectId - Project ID
   * @returns {Promise<Array>} Tasks
   */
  async getByProject(projectId) {
    return await this.taskRepo.findByProject(projectId);
  }

  /**
   * Get tasks for a project grouped by status
   * @param {number} projectId - Project ID
   * @returns {Promise<object>} Tasks grouped by status
   */
  async getByProjectGrouped(projectId) {
    return await this.taskRepo.groupByStatus(projectId);
  }

  /**
   * Create a new task
   * @param {object} taskData - Task data
   * @returns {Promise<object>} Created task
   */
  async create(taskData) {
    // Verify project exists
    const projectExists = await this.projectRepo.exists(taskData.project_id);
    if (!projectExists) {
      const error = new Error('Project not found');
      error.statusCode = 400;
      throw error;
    }
    return await this.taskRepo.create(taskData);
  }

  /**
   * Update a task
   * @param {number} taskId - Task ID
   * @param {object} taskData - Task data to update
   * @returns {Promise<object|null>} Updated task or null
   */
  async update(taskId, taskData) {
    return await this.taskRepo.update(taskId, taskData);
  }

  /**
   * Update task status
   * @param {number} taskId - Task ID
   * @param {string} status - New status
   * @returns {Promise<object|null>} Updated task or null
   */
  async updateStatus(taskId, status) {
    return await this.taskRepo.update(taskId, { status });
  }

  /**
   * Delete a task
   * @param {number} taskId - Task ID
   * @returns {Promise<boolean>} True if deleted
   */
  async delete(taskId) {
    return await this.taskRepo.delete(taskId);
  }

  /**
   * Check if task exists
   * @param {number} taskId - Task ID
   * @returns {Promise<boolean>} True if exists
   */
  async exists(taskId) {
    return await this.taskRepo.findById(taskId) !== null;
  }
}

module.exports = { TaskService };
